### ISSUE TITLE
Character count's count function should be customisable

### ISSUE THREAD

**Issue by 36degrees** [alphagov/govuk-frontend#1364]
Following on from #1104, whatever 'default' way we use to count characters (code points, bytes, characters) ideally it should match the way that a service's backend validation will count them. For this reason, we should consider making the count function customisable.

For example,

```js
new CharacterCount($characterCount).init({ countFunction: string => string.length })
```

**Comment by fofr** [comment]
Same is probably true for word count: https://github.com/alphagov/govuk-design-system-backlog/issues/67#issuecomment-420603632

**Comment by 36degrees** [comment]
Thinking about it, we _might_ also want to support passing a regex as a string, as well as passing a function.

**Comment by romaricpascal** [comment]
Here's a possible plan for implementing this.

API wise, customising the count needs two options:
1. One for the function (the `countFunction` proposed above)
2. One for the maximum (we could reuse `maxlength`, but a new `maximum` option would make the naming more generic)

We also need to consider that options can be passed to the component not only as JavaScript options (where we can have an actual `Function`), but from data-* attributes on the component's root element (where we're limited to string).

This means we need to plan a little map to translate `data-count-function="words"` into an actual function that counts words in the component's constructor.

As a rough idea:
1. Refactor the component's internal behaviour so that setting `maxlength` or `maxwords` stores a `(string) => number` function  for counting in a property (for ex. `getCount`) on the component instance for counting characters or words. That function would then be called in `updateCount` [instead of branching with an `if`](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/src/govuk/components/character-count/character-count.mjs#L348-L352). This prepares the component for running an arbitrary function. Best structure thsed functions in a `{ [string] : (string) => number }` object (see 3.)
2. Allow an arbitrary `(string) => number` function to be passed as JavaScript option (for ex. `countFunction`), which if present gets assigned to the property created in (1). Couple of things to consider there:
	1. Should the component use `maxlength` as the limit when that function is present or a new option (for ex. `maximum`)
	2. Given we accept arbitrary code, what happens if the function throws an error?
	3. Is it OK to limit to synchronous functions or do we need support for asynchronous functions (not that we'd want to make requests to a server for counting, but are there asynchronous browser APIs that may be useful?)
3. Allow `countFunction` to receive a string in addition to a function. When receiving a string, look up in a `{ [string] : (string) => number }` object to translate the string into a function.
4. Deprecate `maxwords` in favour of `countFunction="words"` (and if chosing a new `maximum` option in 2.1, deprecate `maxlength` in favour of `countFunction="characters"`).
5. [Optional] Allow users to register new count functions to support aligning how the component counts with how their backend counts and use data-* attributes for configuring their component.
	1. Weigh between opening access to the `{ [string] : (string) => number }` object directly (for ex. setting it as a static `CharacterCount.countFunctions`) or having a `registerCountFunction(fn)` method of the class or function in the module.

**Comment by colinrotherham** [comment]
I'm taking a look at this over in NHS.UK frontend, and we'd be more than happy to contribute back

There's actually an internal `countType` that fits becoming an option and/or `data-count-type="words"` attribute:

```patch
  {{ govukCharacterCount({
    label: {
      text: "Can you provide more detail?"
    },
    name: "more-detail",
-   maxwords: 200
+   maxlength: 200,
+   countType: "words"
  }) }}
```

It would [carry on working](https://github.com/alphagov/govuk-frontend/blob/v6.1.0/packages/govuk-frontend/src/govuk/components/character-count/character-count.mjs#L365-L366) just fine, with these lines needing an update:

```mjs
const countType = this.config.maxwords ? 'words' : 'characters'
return this.formatCountMessage(remainingNumber, countType)
```

Especially since the `countType` is used to pick i18n strings:

```mjs
this.i18n.t(`${countType}AtLimit`)
this.i18n.t(`${countType}${translationKeySuffix}`)
```

As long as `CharacterCount.countFunctions` supported `"characters"`, `"words"` and `"graphemes"` we don't yet see any user need for customisable count functions

I'll open a PR at our end hopefully this week

**Comment by colinrotherham** [comment]
I've split this into two PRs at our end:

* https://github.com/nhsuk/nhsuk-frontend/pull/1893
* https://github.com/nhsuk/nhsuk-frontend/pull/1895

The count types must be `"characters"` or `"words"`

See if it's something you're interested in

**Comment by colinrotherham** [comment]
We've identified a need for multi-byte string counting where native grapheme support isn't yet available

So to add support for `countFunction` see these two further PRs:

* https://github.com/nhsuk/nhsuk-frontend/pull/1896
* https://github.com/nhsuk/nhsuk-frontend/pull/1897

Plus some other reasons services might need it:

* Line length differences due to `\n` versus `\r\n`
* Word counts that vary based on empty spaces and punctuations
* Trimming empty space before counting

I'll open up discussions with one of our Python + Jinja teams where the `| length` filter is multi-byte (unlike JavaScript)

**Comment by colinrotherham** [comment]
The only other question is how we could avoid (future) breaking changes?

For example, since `countType` is a new option we could make:

1. `countType: "length"` (code points) the default as reported in [#1104](https://github.com/alphagov/govuk-frontend/issues/1104)
2. `countType: "characters"` opt in to `Intl.Segmenter`
3. `countType: "words"` (+ `maxlength`) become the new `maxwords`

Therefore avoiding a migration or removal of `countType: "graphemes"` in future

**Update:** Knowing that [**Intl.Segmenter** also supports `granularity: "word"`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter), would you want both `"characters"` and `"words"` to use the segment length if available?

For example, consider this phrase:

>My mother-in-law—Wait, what?

It matches only 3 words currently:

```js
["My", "mother-in-law—Wait,", "what?"]
```

Yet using the segmenter with `granularity: "word"` it matches 6 words:

```js
["My", "mother", "in", "law", "Wait", "what"]
```

This is because the [Unicode **Default Word Boundary Specification**](https://unicode.org/reports/tr29/#Default_Word_Boundaries) differs quite significantly from `/\S+/g`

**Comment by colinrotherham** [comment]
>Thinking about it, we _might_ also want to support passing a regex as a string, as well as passing a function.

Regarding a word separator regex, you could use [Unicode character class escapes](https://caniuse.com/mdn-javascript_regular_expressions_unicode_character_class_escape) (via the `u` flag) to add the same named groups used internally by [**Intl.Segmenter** with `granularity: "word"`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter) but for older browsers.

```patch
+ const pattern =
+   /[\p{White_Space}\p{Dash_Punctuation}\p{Other_Punctuation}\p{Emoji}\p{Join_Control}]+/gu
+
- const tokens = text.match(/\S+/g) ?? [] // Matches consecutive non-whitespace chars
+ const tokens = text.split(pattern).filter(Boolean)
  this.count = tokens.length
```

Although some browsers shipped `type="module"` before character class escapes were added:

* Chrome 61–63
* Firefox 60–77
* Safari 11

But using try/catch with the RegExp constructor works to avoid errors:

```js
let pattern

try {
  pattern = new RegExp(
    '[\\p{White_Space}\\p{Dash_Punctuation}\\p{Other_Punctuation}\\p{Emoji}\\p{Join_Control}]+',
    'gu'
  )
} catch {
  pattern = /[\s\-‑–—.,;:!\\/]+/g
}
```

Where the regex `/[\s\-‑–—.,;:!\\/]+/g` serves as a minimal fallback

**Comment by colinrotherham** [comment]
We're going to go ahead with the `countType` option used to opt in to `Intl.Segmenter`

* `countType: "length"` (default) continues to [count code points](https://github.com/alphagov/govuk-frontend/issues/1104)
* `countType: "characters"` counts graphemes (user-perceived characters)
* `countType: "words"` counts words regardless of punctuation

Since grapheme clusters are user-perceived characters, it makes sense for `"characters"` to meet this definition

Although we'll check for `maxwords` (deprecated) to opt out and count words by consecutive whitespace only

**Comment by romaricpascal** [comment]
#6995 proposed a `countFunction` option in parallel of the `countType` option it introduces. We'd like to explore if there's an alternative API that would:
- use a single option
- not require the `countFunction` to receive the component's `config` in addition to the `segmenter`.

@colinrotherham highlighted the following constraints:

- [Count type is necessary to get the count message](https://github.com/alphagov/govuk-frontend/blob/v6.2.0/packages/govuk-frontend/src/govuk/components/character-count/character-count.mjs#L365-L366)
- [Count type is necessary to format the count message via i18n](https://github.com/alphagov/govuk-frontend/blob/v6.2.0/packages/govuk-frontend/src/govuk/components/character-count/character-count.mjs#L379-L388)
- [Count type is optional only when ](https://github.com/alphagov/govuk-frontend/blob/v6.2.0/packages/govuk-frontend/src/govuk/components/character-count/template.njk#L6-L15)textareaDescriptionText[ is provided](https://github.com/alphagov/govuk-frontend/blob/v6.2.0/packages/govuk-frontend/src/govuk/components/character-count/template.njk#L6-L15) ([see PR](https://github.com/alphagov/govuk-frontend/pull/2915))

And problems we'll need to solve outside of the shape of the API we provide to users:
- in absence of two options for the component, how to pick the correct message to display to users?
- in absence of the `config` received by the `countFunction`, how would users know `maxwords` has been used asking words to be counted without segmenter
- solution needs to work with the various way to configure the component, particularly `initAll`

**Comment by romaricpascal** [comment]
@colinrotherham proposed an alternative to two options in the form of an object:

```
countType: {
  extends: 'characters',
  updateCount(text, context) {
    return text.length
  }
}
```

**Comment by colinrotherham** [comment]
Thanks @romaricpascal

The downside to my own suggestion there is the `initAll()` problem, where all options apply service-wide

i.e. How would a service-wide function know if it's counting characters or words?

Thankfully the [API design for GOV.UK Frontend takes `data-attributes`](https://frontend.design-system.service.gov.uk/configure-components/) as higher priority over JavaScript configuration, which is why a separate `countType` option can still be set in HTML and magically picked up inside the count function 🪄 

```mjs
import { importAll, CharacterCount } from 'govuk-frontend'

initAll({
  characterCount: {
    countFunction(text, context) {
      // Q: Are we counting characters or words here?
      // A: See `context.config.countType` from `data-count-type`
      return text.length
    }
  }
})
```
